File: python/tests/string_concat.py
import time, json, tracemalloc, gc

class StringConcatTest:
    name = "StringConcat"

    def __init__(self, n=500_000):
        self.n = n

    def concat_plus(self):
        s = ""
        for _ in range(self.n):
            s += "a"
        return len(s)

    def concat_join(self):
        return len("".join(["a"] * self.n))

    def run_once(self):
        gc.collect()
        tracemalloc.start()
        t0 = time.perf_counter()

        length_plus = self.concat_plus()
        t1 = time.perf_counter()

        length_join = self.concat_join()
        t2 = time.perf_counter()

        current, peak = tracemalloc.get_traced_memory()
        tracemalloc.stop()

        plus_time = t1 - t0
        join_time = t2 - t1
        total_time = t2 - t0

        return {
            "plus_time_s": plus_time,
            "join_time_s": join_time,
            "total_time_s": total_time,
            "mem_peak_bytes": peak,
            "length_plus": length_plus,
            "length_join": length_join,
            "ops_per_sec_plus": self.n / plus_time if plus_time > 0 else float("inf"),
            "ops_per_sec_join": self.n / join_time if join_time > 0 else float("inf")
        }

    def run(self, runs=3):
        results = [self.run_once() for _ in range(runs)]
        total_times = sorted(r["total_time_s"] for r in results)
        plus_times = sorted(r["plus_time_s"] for r in results)
        join_times = sorted(r["join_time_s"] for r in results)
        peaks = sorted(r["mem_peak_bytes"] for r in results)
        mid = len(results) // 2

        return {
            "name": self.name,
            "runs": runs,
            "median_total_time_s": total_times[mid],
            "median_plus_time_s": plus_times[mid],
            "median_join_time_s": join_times[mid],
            "median_peak_mem_bytes": peaks[mid],
            "raw": results
        }

if __name__ == "__main__":
    result = StringConcatTest(500_000).run()

File: ruby/tests/string_concat.rb
require "json"
require "time"

class StringConcatTest
  attr_reader :name

  def initialize(n = 50_000)
    @name = "StringConcat"
    @n = n
  end

  def concat_plus
    s = ""
    @n.times { s += "a" }
    s.length
  end

  def concat_join
    Array.new(@n, "a").join.length
  end

  def run_once
    t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    len_plus = concat_plus
    t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    len_join = concat_join
    t2 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    {
      "plus_time_s" => t1 - t0,
      "join_time_s" => t2 - t1,
      "total_time_s" => t2 - t0,
      "length_plus" => len_plus,
      "length_join" => len_join
    }
  rescue Interrupt
    { "error" => "Interrupted during test" }
  end

  def run(runs = 3, _initialize = nil)
    results = Array.new(runs) { run_once }
    valid_results = results.reject { |r| r.key?("error") }

    if valid_results.empty?
      return { "name" => @name, "error" => "All runs interrupted" }
    end

    total = valid_results.map { |r| r["total_time_s"] }.sort
    plus = valid_results.map { |r| r["plus_time_s"] }.sort
    join = valid_results.map { |r| r["join_time_s"] }.sort
    median = valid_results.size / 2

    {
      "name" => @name,
      "runs" => runs,
      "median_total_time_s" => total[median],
      "median_plus_time_s" => plus[median],
      "median_join_time_s" => join[median],
      "raw" => results
    }
  end
end

if __FILE__ == $0
  test = StringConcatTest.new(50_000)
  result = test.run
  timestamp = Time.now.strftime('%Y%m%d_%H%M%S')
  filename = "results_ruby_string_concat_#{timestamp}.json"
  File.write(filename, JSON.pretty_generate(result))
  puts "Results saved to #{filename}"
  puts JSON.pretty_generate(result)
end